### ISSUE TITLE
Const type parameter behaves differently when using inline func vs func declaration

### ISSUE THREAD

**Issue by kovalenko0** [microsoft/TypeScript#55428]
### 🔎 Search Terms

const type parameter

### 🕗 Version & Regression Information

Replicates on 5.0.5, 5.1.6 and nightly (5.3.0-dev.20230817)

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.1.6#code/MYewdgzgLgBArhApgJwgWQIYAcBcNNYA80yAlmAOYA0MA3jGBgLaJ4nnUwYWsNxMAjFDAC+APhgBeBogDu+bAAoAlAChQkWNEQA3RAEYpMJtgD6PKAHlkAYWSIMURIoQp02GgHJtezzRVSEsD2jogAqkjIit5OvsrKMAD0iTCIAB5YiMBOACZcYKnIyCDIMJ4ACsWZyFAAnmXciJ4wpBDGrRAcLQV1mTAAdIOe6uDQMD6IAExGJljmiFa2IU4uke5YXhN+MAGSEor0jCx4MbpNovFJKbIlANZtGG3pmdmIOaqqAGZwYNmk4DBgg4nBEUIojrx2JQErRVDAYPYoHBkAVDsxEKJVCIPt9flB-gVZvNFnZgYhCABpGgaMYANTEilmeAIlJo9Jot0QtTwVMBy14uwktOUeFpdDhLU+OwAhLN+gALR6KTm1S5ypBQZVc6n8lRqeGI5GE7D9Cxa1VcNq0rFAA

### 💻 Code

```ts
const usersMap: Map<string, { name: string, age: number }> = new Map()
const steve1 = map_getOrCreate(usersMap, 'steve', () => createUser('steve')) // expected an error 'Property 'age' is missing in type ...'
const steve2 = map_getOrCreate(usersMap, 'steve', () => ({ name: 'steve' })) // works as expected

function createUser(name: string) {
  return { name }
}

function map_getOrCreate<K, const V>(map: Map<K, V>, key: K, create: () => V): V {
  if (!map.has(key)) map.set(key, create())
  return map.get(key) as V
}
```

### 🙁 Actual behavior

Call accepts a function that returns an incompatible type
```
 map_getOrCreate(usersMap, 'steve', () => createUser('steve') /* returns { name: string } */ )
```

### 🙂 Expected behavior

Expected an error `Property 'age' is missing in type ...`

**Comment by Andarist** [comment]
Those 2 calls are not the same. An extra function introduces different semantics and "outer" `const` doesn't have an effect on the inner call. If we make them more alike then it works as you expected: [TS playground](https://www.typescriptlang.org/play?ts=5.3.0-dev.20230818#code/MYewdgzgLgBArhApgJwgWQIYAcBcNNYA80yAlmAOYA0MA3jGBgLaJ4nkUDcMGFrDcJgCMUMAL4A+GAF4GiAO75sACgCUnAFChIsaIgBuiAIwyYTbAH0+UAPLIAwskQYoiZQhTpsNAER7DPjRqMlLATi6IAKpIyMp+rgGq6lrg0DD+iABMpuZYVoi2DuGu7jFeWL4ZgTDB0lLK9IwsePEGiD7iSZoaAGZwYMBQpOAwYc6u0SiEACowiAAermAAJhDpUGSUEspN-NOqdBowME5QcMhgdAzMiOI8a9rQmmIavf2Dw5e5+YWO44iEADSNEesAAatsjmZsHgCECaBCqFCANaIACeeGBULGETwtSkYI0qjwYMOx1IPRqAEJcgA6AAWGAgylRaKS0KwtKQUBZ6JBxTcXShp3OX2wtOsvLZ9xgYOeGiAA)

```ts
const usersMap: Map<string, { name: string; age: number }> = new Map();
const steve1 = map_getOrCreate(usersMap, "steve", () => createUser("steve")); // error
const steve2 = map_getOrCreate(usersMap, "steve", () => ({ name: "steve" })); // error

function createUser<T extends string>(name: T) {
  return { name } as const;
}

function map_getOrCreate<K, const V>(
  map: Map<K, V>,
  key: K,
  create: () => V
): V {
  if (!map.has(key)) map.set(key, create());
  return map.get(key) as V;
}
```

Before this change the variant that errors calls `getCommonSupertype(baseCandidates)` in `getInferredType` with `baseCandidates` equals to `['{ name: string; age: number; }', '{ readonly name: "steve"; }']`. The one that doesn't error does the same with `['{ name: string; age: number; }', '{ name: string; }']` and it seems that `getCommonSupertype` handles both of those differently.

**Comment by RyanCavanaugh** [comment]
There's no error here because the inferred type of `steve1` is different from the inferred type of `steve2`, so the program is still sound
